### ISSUE TITLE
gui wx incompatibilities with new wxpython 'phoenix'

### ISSUE THREAD

**Issue by karlacio** [ipython/ipython#10055]
Hi,

running a program that uses wxpython gui with iphyton:
`ipython --gui wx --i my_program.py`

gives an error like:

> File ".../site-packages/IPython/terminal/pt_inputhooks/wx.py", line 76, in inputhook_wx2
>     assert wx.Thread_IsMain()
> AttributeError: 'module' object has no attribute 'Thread_IsMain'
> 

On wxpython 'phoenix' seems that the method `Thread_IsMain` as been replaced by `IsMainThread`.

Replacing it on wx.py it seems to work.

I'm running

> Python 2.7.10 
> IPython 5.1.0
> wx.__version__ 3.0.3.dev2553+47f1fd2
> on OSX 10.11.6

**Comment by takluyver** [comment]
Oh, wx is going again? I'd recommend making a new event loop integration for it. It looks like it doesn't have a clearly distinct version number, so we can't call it e.g. `wx3`. Maybe `wxphoenix`?

**Comment by jaredl7** [comment]
I'm using IPython 6.2.1 and I had the same issue as @karlacio.

Using the example from here: https://github.com/ipython/ipython/blob/master/examples/IPython%20Kernel/gui/gui-wx.py

and, applying the fix that @karlacio recommended (this is line 118 in `inputhook_wx3` in IPython 6.2.1) - i.e.:

**`assert wx.Thread_IsMain()`** -> **`assert wx.IsMainThread()`**

one obtains a new error: 

> AttributeError: module 'wx' has no attribute 'Timer'

which appears to be solved by updating 

**`app.ProcessIdle()`** to **`evtloop.ProcessIdle()`** 

on **`<Python version>/site-packages/IPython/lib/inputhookwx.py`** and in **`<Python version>/site-packages/IPython/terminal/pt_inputhooks/wx.py`**

Having referenced this updated sample from wxPython: https://github.com/wxWidgets/wxPython/blob/4b20d984aaaf4a5d7dfcdc1490c0b42489d452f7/samples/mainloop/mainloop.py

Applying these fixes solves the issue, and one is able to run the example cited without issue.

**Comment by takluyver** [comment]
Do you want to make a new inputhook for wxpython 4 with these changes? I.e. add `IPython.terminal.pt_inputhooks.wx4`. We don't use the `IPython.lib.inputhook*` ones any more.

There's also a corresponding set of event loop integration pieces in ipykernel, which may need a `wx4` option: https://github.com/ipython/ipykernel/blob/2baf8160b23bb4a0678da2033099eff9423316ae/ipykernel/eventloops.py#L144

Or if you're confident that these changes won't break things for people still using wxpython 3, you could make the changes to the code in place.

**Comment by jaredl7** [comment]
Sure @takluyver, I can give it a go! While I'm reasonably confident that it shouldn't break things, I'll be more comfortable testing it for `wx3` and `wx4` so as to ensure that it doesn't break anything for anyone. I'll ping back when I've verified both.

Thanks!

**Comment by rdemaria** [comment]
In my case I had also a problem with a deprecation warning
```
IPython/terminal/pt_inputhooks/wx.py:106: wxPyDeprecationWarning: Using deprecated class EventLoop. Use GUIEventLoop instead.
  evtloop = wx.EventLoop()
```

It seems I fixed my setup with 
```
cd site-packages/IPython
sed -i "s/EventLoop(/GUIEventLoop(/" lib/inputhookwx.py terminal/pt_inputhooks/wx.py
sed -i "s/app.ProcessIdle/evtloop.ProcessIdle/" lib/inputhookwx.py terminal/pt_inputhooks/wx.py
sed -i "s/Thread_IsMain/IsMainThread/" lib/inputhookwx.py terminal/pt_inputhooks/wx.py
```

**Comment by dhomeier** [comment]
Got the same `AttributeError`s and `wxPyDeprecationWarning` with IPython 7.5.0, which still seems to use a single hook for wx3 and wx4. The switch to `IsMainThread` apparently dates back to [2012](https://github.com/wxWidgets/Phoenix/commit/a1163deb783995f1dde89c4f6d00f7ddce8a2a87#diff-5a6bec56100c9dd4b2e5542bd752c4ce).
So maybe a quick fix (workaround?)  for that one could be
```
if int(wx.__version__[0]) > 3: wx.Thread_IsMain = wx.IsMainThread
```
(or even `if not hasattr(wx, 'Thread_IsMain'):`
Not sure how to properly test for the `ProcessIdle` instance ...

With the above fixes on exiting an `ipython --pylab=wx` session I am still getting this
```
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/sw/lib/python3.5/site-packages/matplotlib/_pylab_helpers.py", line 74, in destroy_all
    manager.destroy()
  File "/sw/lib/python3.5/site-packages/matplotlib/backends/backend_wx.py", line 1287, in destroy
    self.frame.Destroy()
  File "/sw/lib/python3.5/site-packages/matplotlib/backends/backend_wx.py", line 1246, in Destroy
    if not self.IsBeingDeleted():
RuntimeError: wrapped C/C++ object of type FigureFrameWxAgg has been deleted
```
but the ipython kernel exits, and anyway that appears more of a matplotlib issue.

**Comment by tacaswell** [comment]
The relationship between Matplotlib, IPython, and the GUI frameworks is a bit mixed.  Matplotlib owns the widget code (because we own the plots) and IPython owns the input hooks (because they own the terminal).

IIRC, the issue with the RuntimeError on exit is due to race conditions where we try to tear the c++ objects down twice.    It looks like we are trying to avoid that by checking if the window is being torn down (https://wxpython.org/Phoenix/docs/html/wx.Window.html#wx.Window.IsBeingDeleted), but apparently are still losing the race.  @dhomeier Could you put in a PR to Matplotlib to either catch and ignore that error or is there some better check we should be doing to avoid his race condition?

**Comment by dhomeier** [comment]
@tacaswell I can file an issue; not sure that I understand enough of what's going on there to file an acceptable PR. Simply inserting a `if self and not self.IsBeingDeleted():` gets me rid of the RuntimeError, but it doesn't look very robust to me...

**Comment by Carreau** [comment]
closed by #11911
